pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -26,10 +36,177 @@ def synchronize(remote_pk, repository_pk):
+            cr.xml_parse_filelists(filelists_xml_path, newpkgcb=newpkgcb)
+            cr.xml_parse_other(other_xml_path, newpkgcb=newpkgcb)
+            return packages
+
+        async def parse_updateinfo(updateinfo_xml_path):
+            """
+            Parse updateinfo.xml to extact update info.
+
+            Args:
+                updateinfo_xml_path: a path to a downloaded updateinfo.xml
+
+            Returns:
+                :obj:`list` of :obj:`createrepo_c.UpdateRecord`: parsed update records
+
+            """
+            uinfo = cr.UpdateInfo()
+
+            # TODO: handle parsing errors/warnings, warningcb callback can be used
+            cr.xml_parse_updateinfo(updateinfo_xml_path, uinfo)
+            return uinfo.updates
+
+        with ProgressBar(message='Downloading and Parsing Metadata') as pb:
+            downloader = self.remote.get_downloader(urljoin(self.remote.url,
+                                                            'repodata/repomd.xml'))
+            # TODO: decide how to distinguish between a mirror list and a normal repo
+            result = await downloader.run()
+            pb.increment()
+
+            repomd_path = result.path
+            repomd = cr.Repomd(repomd_path)
+            package_repodata_urls = {}
+            downloaders = []
+
+            for record in repomd.records:
+                if record.type in PACKAGE_REPODATA:
+                    package_repodata_urls[record.type] = urljoin(self.remote.url,
+                                                                 record.location_href)
+                elif record.type in UPDATE_REPODATA:
+                    updateinfo_url = urljoin(self.remote.url, record.location_href)
+                    downloader = self.remote.get_downloader(updateinfo_url)
+                    downloaders.append([downloader.run()])
+                else:
+                    log.info(_('Unknown repodata type: {t}. Skipped.').format(t=record.type))
+                    # TODO: skip databases, save unknown types to publish them as-is
+
+            # to preserve order, downloaders are created after all repodata urls are identified
+            package_repodata_downloaders = []
+            for repodata_type in PACKAGE_REPODATA:
+                downloader = self.remote.get_downloader(package_repodata_urls[repodata_type])
+                package_repodata_downloaders.append(downloader.run())
+
+            downloaders.append(package_repodata_downloaders)
+
+            # asyncio.gather is used to preserve the order of results for package repodata
+            pending = [asyncio.gather(*downloaders_group) for downloaders_group in downloaders]
+
+            while pending:
+                done, pending = await asyncio.wait(pending, return_when=asyncio.FIRST_COMPLETED)
+                for downloader in done:
+                    results = downloader.result()
+                    if results[0].url == package_repodata_urls['primary']:
```

Comment:
Here it would be nice to have a way to identify a coroutine I have results from (this is my question#3 from [the long comment](https://github.com/pulp/pulp_rpm/pull/1160#issuecomment-414135528)). 
Currently I compare URLs. Is there a better way to do this?

Reply:
When I want to send "generic info" along with the result of some co-routine I have this little hack I made. I use it here:  https://github.com/pulp/pulp/blob/master/plugin/pulpcore/plugin/stages/artifact_stages.py#L190-L191

The trick here is to make a co-routine whose only point is to hand you back data you passed into it. Then you join this "result" data to your original co-routine of interest by using `asyncio.gather(dummy_coroutine, the_coroutine_of_interest)` and await on the gathered result (the coroutine gather creates). The result of this coroutine will be a 2-position iterable with the first position being the data you originally passed to the dummy coroutine and the result of your actual downloader coroutine.